From fd7cc127b364d5571f936f130f6da01d4ea79201 Mon Sep 17 00:00:00 2001
From: Kensuke Ueda <uedake@nttdocomo.co.jp>
Date: Thu, 8 Aug 2013 19:00:48 +0900
Subject: [PATCH 13/20] Improve a calculation method for LTE antenna reception
 level

This proposal will fix a problem that Android devices could
show different levels of antenna bars under the same radio
environment if each device uses different modem.
This behavior makes confusion toward users and others.

That behavior comes from using RSSNR as a metric for
determining antenna bars. Unfortunately, RSSNR does NOT define
performance requirements and error tolerance level in 3GPP.
Therefore, accuracy of RSSNR measurement could be inconsistent
among Android devices or modems.

The patch we would like to contribute here will be offering
Android device to use RSRQ for antenna bar metric instead of RSSNR.

1.)RSRQ, like RSSNR, takes interference into account
   when measuring the radio quality.
   Android already supports measurement of RSRQ.

2.)The performance requirement for RSRQ is already defined in 3GPP.

3.)DOCOMO and other many mobile operators use RSRP and RSRQ
   when designing LTE coverage areas, and alignment here will
   be very beneficial for both users and operators.

Hope this patch helps to improve Android antenna bar accuracy and consistency today.

Change-Id: Ieaaeffad125e5c0aaea14ce4a0482adfe8da5313
---
 core/res/res/values/lineage_config.xml             |  8 ++
 core/res/res/values/lineage_symbols.xml            |  4 +
 .../java/android/telephony/SignalStrength.java     | 90 +++++++++++++++-------
 3 files changed, 75 insertions(+), 27 deletions(-)

diff --git a/core/res/res/values/lineage_config.xml b/core/res/res/values/lineage_config.xml
index 684d121..3086d5b 100644
--- a/core/res/res/values/lineage_config.xml
+++ b/core/res/res/values/lineage_config.xml
@@ -15,6 +15,14 @@
      limitations under the License.
 -->
 <resources>
+    <!-- Flag indicating whether strict threshold is used, or lenient threshold is used,
+          when evaluating RSRP for LTE antenna bar display
+           0. Strict threshold
+           1. Lenient threshold
+    -->
+    <integer name="config_LTE_antenna_bar_quantity_metric">0</integer>
+    <integer name="config_LTE_RSRP_threshold_type">0</integer>
+
     <!-- Button backlight -->
     <integer name="config_buttonBrightnessSettingDefault">255</integer>
     <bool name="config_deviceHasVariableButtonBrightness">false</bool>
diff --git a/core/res/res/values/lineage_symbols.xml b/core/res/res/values/lineage_symbols.xml
index 793ad1b..035323f 100644
--- a/core/res/res/values/lineage_symbols.xml
+++ b/core/res/res/values/lineage_symbols.xml
@@ -15,6 +15,10 @@
      limitations under the License.
 -->
 <resources>
+    <!-- From SignalStrength -->
+    <java-symbol type="integer" name="config_LTE_antenna_bar_quantity_metric" />
+    <java-symbol type="integer" name="config_LTE_RSRP_threshold_type" />
+
     <!-- ADB notification -->
     <java-symbol type="string" name="adb_net_active_notification_title" />
     <java-symbol type="string" name="adb_both_active_notification_title" />
diff --git a/telephony/java/android/telephony/SignalStrength.java b/telephony/java/android/telephony/SignalStrength.java
index 4e56396..70c923a 100644
--- a/telephony/java/android/telephony/SignalStrength.java
+++ b/telephony/java/android/telephony/SignalStrength.java
@@ -854,6 +854,7 @@ public class SignalStrength implements Parcelable {
          * RSRQ = quality of signal dB = Number of Resource blocks*RSRP/RSSI
          * SNR = gain = signal/noise ratio = -10log P1/P2 dB
          */
+        int rsrqIconLevel = -1;
         int rssiIconLevel = SIGNAL_STRENGTH_NONE_OR_UNKNOWN, rsrpIconLevel = -1, snrIconLevel = -1;
 
         if (mLteRsrp > MAX_LTE_RSRP || mLteRsrp < MIN_LTE_RSRP) {
@@ -879,37 +880,72 @@ public class SignalStrength implements Parcelable {
             }
         }
 
-        /*
-         * Values are -200 dB to +300 (SNR*10dB) RS_SNR >= 13.0 dB =>4 bars 4.5
-         * dB <= RS_SNR < 13.0 dB => 3 bars 1.0 dB <= RS_SNR < 4.5 dB => 2 bars
-         * -3.0 dB <= RS_SNR < 1.0 dB 1 bar RS_SNR < -3.0 dB/No Service Antenna
-         * Icon Only
-         */
-        if (mLteRssnr > 300) snrIconLevel = -1;
-        else if (mLteRssnr >= 130) snrIconLevel = SIGNAL_STRENGTH_GREAT;
-        else if (mLteRssnr >= 45) snrIconLevel = SIGNAL_STRENGTH_GOOD;
-        else if (mLteRssnr >= 10) snrIconLevel = SIGNAL_STRENGTH_MODERATE;
-        else if (mLteRssnr >= -30) snrIconLevel = SIGNAL_STRENGTH_POOR;
-        else if (mLteRssnr >= -200)
-            snrIconLevel = SIGNAL_STRENGTH_NONE_OR_UNKNOWN;
-
-        if (DBG) log("getLTELevel - rsrp:" + mLteRsrp + " snr:" + mLteRssnr + " rsrpIconLevel:"
-                + rsrpIconLevel + " snrIconLevel:" + snrIconLevel
-                + " lteRsrpBoost:" + mLteRsrpBoost);
-
-        /* Choose a measurement type to use for notification */
-        if (snrIconLevel != -1 && rsrpIconLevel != -1) {
+        int secondMetricType = Resources.getSystem().getInteger(com.android.internal.R.integer.config_LTE_antenna_bar_quantity_metric);
+        if (secondMetricType == 0){
             /*
-             * The number of bars displayed shall be the smaller of the bars
-             * associated with LTE RSRP and the bars associated with the LTE
-             * RS_SNR
+             * Values are -200 dB to +300 (SNR*10dB) RS_SNR >= 13.0 dB =>4 bars 4.5
+             * dB <= RS_SNR < 13.0 dB => 3 bars 1.0 dB <= RS_SNR < 4.5 dB => 2 bars
+             * -3.0 dB <= RS_SNR < 1.0 dB 1 bar RS_SNR < -3.0 dB/No Service Antenna
+             * Icon Only
              */
-            return (rsrpIconLevel < snrIconLevel ? rsrpIconLevel : snrIconLevel);
-        }
+            if (mLteRssnr > 300) snrIconLevel = -1;
+            else if (mLteRssnr >= 130) snrIconLevel = SIGNAL_STRENGTH_GREAT;
+            else if (mLteRssnr >= 45) snrIconLevel = SIGNAL_STRENGTH_GOOD;
+            else if (mLteRssnr >= 10) snrIconLevel = SIGNAL_STRENGTH_MODERATE;
+            else if (mLteRssnr >= -30) snrIconLevel = SIGNAL_STRENGTH_POOR;
+            else if (mLteRssnr >= -200)
+                snrIconLevel = SIGNAL_STRENGTH_NONE_OR_UNKNOWN;
+
+            if (DBG) log("getLTELevel - rsrp:" + mLteRsrp + " snr:" + mLteRssnr + " rsrpIconLevel:"
+                    + rsrpIconLevel + " snrIconLevel:" + snrIconLevel
+                    + " lteRsrpBoost:" + mLteRsrpBoost);
+
+            /* Choose a measurement type to use for notification */
+            if (snrIconLevel != -1 && rsrpIconLevel != -1) {
+                /*
+                 * The number of bars displayed shall be the smaller of the bars
+                 * associated with LTE RSRP and the bars associated with the LTE
+                 * RS_SNR
+                 */
+                return (rsrpIconLevel < snrIconLevel ? rsrpIconLevel : snrIconLevel);
+            }
+
+            if (snrIconLevel != -1) return snrIconLevel;
+
+            if (rsrpIconLevel != -1) return rsrpIconLevel;
+       }
+       else if (secondMetricType == 1){
+            /*
+             * Values are -34 dB to -3dB RSRQ >= -12dB =>4 bars -14
+             * dB <= RSRQ < -12 dB => 3 bars -17 dB <= RSRQ < -14 dB => 2 bars
+             * -19 dB <= RSRQ < -17 dB 1 bar RS_SNR < -34 dB/No Service Antenna
+             * Icon Only
+             */
+            if (mLteRsrq > -3) rsrqIconLevel = -1;
+            else if (mLteRsrq >=  -12) rsrqIconLevel = SIGNAL_STRENGTH_GREAT;
+            else if (mLteRsrq >=  -14) rsrqIconLevel = SIGNAL_STRENGTH_GOOD;
+            else if (mLteRsrq >=  -17) rsrqIconLevel = SIGNAL_STRENGTH_MODERATE;
+            else if (mLteRsrq >=  -19) rsrqIconLevel = SIGNAL_STRENGTH_POOR;
+            else if (mLteRsrq >=  -34)
+                rsrqIconLevel = SIGNAL_STRENGTH_NONE_OR_UNKNOWN;
+
+            if (DBG) log("getLTELevel - rsrp:" + mLteRsrp + " rsrq:" + mLteRsrq + " rsrpIconLevel:"
+                    + rsrpIconLevel + " rsrqIconLevel:" + rsrqIconLevel);
+
+            /* Choose a measurement type to use for notification */
+            if (rsrqIconLevel != -1 && rsrpIconLevel != -1) {
+                /*
+                 * The number of bars displayed shall be the smaller of the bars
+                 * associated with LTE RSRP and the bars associated with the LTE
+                 * RSRQ
+                 */
+                return (rsrpIconLevel < rsrqIconLevel ? rsrpIconLevel : rsrqIconLevel);
+            }
 
-        if (snrIconLevel != -1) return snrIconLevel;
+            if (rsrqIconLevel != -1) return rsrqIconLevel;
 
-        if (rsrpIconLevel != -1) return rsrpIconLevel;
+            if (rsrpIconLevel != -1) return rsrpIconLevel;
+       }
 
         /* Valid values are (0-63, 99) as defined in TS 36.331 */
         // TODO the range here is probably supposed to be (0..31, 99). It's unclear if anyone relies
-- 
2.7.4

